components/usage/pkg/controller/reconciler_test.go

```diff
@@ -14,67 +14,102 @@ import (
 	"time"
 )
 
-func TestUsageReconciler_Reconcile(t *testing.T) {
-	conn := db.ConnectForTests(t)
-	instanceStatus := []byte(`{"phase": "stopped", "conditions": {"deployed": false, "pullingImages": false, "serviceExists": false}}`)
+func TestUsageReconciler_ReconcileTimeRange(t *testing.T) {
+	type Scenario struct {
+		Name    string
+		NowFunc func() time.Time
+
+		Memberships []db.TeamMembership
+		Workspaces  []db.Workspace
+		Instances   []db.WorkspaceInstance
+
+		Expected *UsageReconcileStatus
+	}
+
 	startOfMay := time.Date(2022, 05, 1, 0, 00, 00, 00, time.UTC)
 	startOfJune := time.Date(2022, 06, 1, 0, 00, 00, 00, time.UTC)
-	teamID := uuid.New()
-	userID := uuid.New()
-	membership := db.TeamMembership{
-		ID:     uuid.New(),
-		TeamID: teamID,
-		UserID: userID,
-		Role:   db.TeamMembershipRole_Member,
-	}
-	workspace := dbtest.NewWorkspace(t, db.Workspace{
-		ID:      "gitpodio-gitpod-gyjr82jkfnd",
-		OwnerID: userID,
-	})
-	instances := []db.WorkspaceInstance{
-		// Ran throughout the reconcile period
-		{
-			ID:           uuid.New(),
-			WorkspaceID:  workspace.ID,
-			CreationTime: db.NewVarcharTime(time.Date(2022, 05, 1, 00, 00, 00, 00, time.UTC)),
-			StoppedTime:  db.NewVarcharTime(time.Date(2022, 06, 1, 1, 0, 0, 0, time.UTC)),
-			Status:       instanceStatus,
-		},
-		// Still running
-		{
-			ID:           uuid.New(),
-			WorkspaceID:  workspace.ID,
-			CreationTime: db.NewVarcharTime(time.Date(2022, 05, 30, 00, 00, 00, 00, time.UTC)),
-			Status:       instanceStatus,
-		},
-		// No creation time, invalid record
-		{
-			ID:          uuid.New(),
-			WorkspaceID: workspace.ID,
-			StoppedTime: db.NewVarcharTime(time.Date(2022, 06, 1, 1, 0, 0, 0, time.UTC)),
-			Status:      instanceStatus,
-		},
-	}
+	instanceStatus := []byte(`{"phase": "stopped", "conditions": {"deployed": false, "pullingImages": false, "serviceExists": false}}`)
+
+	scenarios := []Scenario{
+		(func() Scenario {
```

Comment:
Why do we need the `func` here. It just returns one `Scenario` without taking any inputs.

Reply:
To encapsulate the extra variables that are needed for this scenario outside of the scope of other scenarios. I've a follow-up PR which adds more complex tests to this.